JavaScriptCodeConventions.md

```diff
@@ -1,13 +1,79 @@
+
+## Contents
+1. [Introduction](#introduction)
+2. [Structure](#structure)
+3. [Applying these Conventions in your Editor](#applying-these-conventions-in-your-editor)
 
 ## Introduction
 These JavaScript style guidelines are intended to be used by both the R&D and marketing team as part of a broader effort to standardize JavaScript style across the company and as an early facilitator for the fact that Ultimaker continues to offer more web-based products such as the Cura Cloud services and Cura Connect.
 
-## AirBnB as Baseline Style
-To reduce the need to build comprehensive style guidelines ourselves, we have decided to adopt the well-known and widely used **AirBnB style**. The full documentation for AirBnB style [can be found on GitHub](https://github.com/airbnb/javascript). For the most part, this style is already very similar to what is used by front-end teams at Ultimaker with a few differences, which are outlined below. It was also chosen because of it was designed for usage with React.js, which is used by the R&D department. It is applicable to Vue.js as well, however, and can be used by the marketing department.
+## Structure
+Per the convention for `Ultimaker/Meta`, rules within this document override Ultimaker Generic rules (`code_conventions.md`), which override the baseline standard (AirBnB). These can be considered a bottom, middle, and top layer of the standard.
+
+### Bottom Layer: AirBnB
+As a foundation, Ultimaker uses the well-known and widely used **AirBnB style**. The full documentation for AirBnB style [can be found on GitHub](https://github.com/airbnb/javascript).
+
+> **Why?** For the most part, this style is already very similar to what is used by front-end teams at Ultimaker. It was also chosen because of it was designed for usage with React.js, which is used by the R&D department.
+
+### Middle Layer: UM Generic
+In certain places, the UM Generic style (`code_conventions.md`) overrides AirBnB style. These exceptions can be inferred by comparing the two documents, however for clarity they are as follows:
+
+- Use 4 spaces for indentation.
+
+	```
+	Example...
+	```
+	
+- Use 1-2 leading underscores to denote protected and private properties (respectively).
+
+	```
+	Example...
+	```
+
+### Top Layer: UM JavaScript
+In certain cases, the UM JavaScript style overrides the UM Generic style (`code_conventions.md`). These exceptions can be inferred by comparing the two documents however for clarity they are as follows:
+
+- Use camelCase for both variable names and method/function names.
+
+	> **Why?** In JavaScript, functions are variables, and camelCase is effectively industry-standard.
+	
+	```
+	Example...
+	```
+
+- Always put brackets on the same line .
+
+	> **Why?** Readability is improved and same-line is effectively industry-standard.
```

Comment:
Readability is subjective here. I feel it's a bit less readable. The one reason we can agree on that it's industry standard.

Reply:
Of course it's a personal thing but the crux of it is readability (not a technical thing). It's a personal preference which may not apply to everyone but the majority seem to prefer it that way.

Reply:
I think part of it for me too is I'm used to typing `{` in my editor and it then puts a `}` on the other side of my cursor so when I hit enter I start writing in that block. It feels a lot easier (I know it's a tiny difference but when you do it 500 times per day it adds up) than hitting enter before the first `{`.

Reply:
Again; if it's a personal preference it should not be in the code standard as a statement that it does improve it. Eg; "I like vanilla ice cream and the entire team likes it" is not the same as "Vanilla ice cream is the best tasting kind". 

It being a (web) industry standard is something that can be measured, so that is a statement you could make (eg; "We eat vanilla ice cream because it's most readily available" is a good reason, even though it might not be the only reason to do it (Because you also happen to think it's the nicest flavor)).

Reply:
I think the #1 reason to have a code style at all is to promote consistency. If you are defining multiple code styles that are inconsistent with each other, that defeats the purpose of having a code style.

Reply:
I'd prefer it if it was removed because personal opinions don't have a place in technical documentation. 

It's a style, but that does not mean that the argument should be "I felt like it". If that's the case, you should put that why at every choice you make. The factual arguments for this change are pretty decent (There is no personal preference involved in these); 
1. Existing codebases use it (thus; changing it is expensive)
2. It matches with industry standards. (Thus; making it easier to adopt for new employees)

Both of those are factual and challengable reasons (eg; I can validate that this is or is not the case. So if it turns out that this statement is wrong, the decision can be re-evaluated). "I feel its more readable" is not. Especially since I disagree with that it's more readable (thus, already creating the first conflict in that statement). 

If you really want to put it in, it should be "It was put to a vote and these teams found it to be more readable" is correct imho (as this is a statement that can be validated/challenged).